app/controllers/oauth2/authorization_controller.rb

```diff
@@ -0,0 +1,150 @@
+# NOTE: This class uses Rails views in a non-legacy manner.  This is by design.
+# Do not attempt to Emberize this controller without first studying the
+# potential security effects it may have (namely CSRF)
+#
+# WARNING: IF YOU HAVE NOT READ RFC 6749, DO NOT TOUCH THIS.
+#
+# Throughout this file there are references to RFC 6749 in parentheses.  The
+# numbers separated by dots are the section id from the spec, the number after
+# "P" is the paragraph, and "S" is sentence.
+class OAuth2::AuthorizationController < ApplicationController
+  # Valid response errors (4.1.2.1, 4.2.2.1)
+  VALID_ERRORS = %i[invalid_request unauthorized_client access_denied
+                    unsupported_response_type invalid_scope server_error
+                    temporarily_unavailable]
+  # Not specified by RFC 6749, these are internal
+  VALID_FATAL_ERRORS = %i[invalid_client invalid_redirect]
```

Comment:
Dunno if it will impact the implementation rules, but wouldn't be better to move those validations to a lib and throw exceptions that we can handle in the controller? It seems to be getting a little fat with all these private methods and all those RFC comments